modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/snapshot/daemon/SelectablePeerDiscoveryDelay.scala

```diff
@@ -0,0 +1,60 @@
+package org.tessellation.sdk.infrastructure.snapshot.daemon
+
+import cats.effect.Async
+import cats.syntax.applicative._
+import cats.syntax.eq._
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+
+import scala.concurrent.duration.DurationInt
+
+import org.tessellation.cli.AppEnvironment
+import org.tessellation.cli.AppEnvironment.Dev
+import org.tessellation.schema.node.NodeState.Ready
+import org.tessellation.schema.peer.Peer
+import org.tessellation.sdk.domain.cluster.storage.ClusterStorage
+import org.tessellation.sdk.domain.snapshot.PeerDiscoveryDelay
+
+import eu.timepit.refined.auto._
+import eu.timepit.refined.types.numeric.PosInt
+import org.typelevel.log4cats.slf4j.Slf4jLogger
+import retry.RetryPolicies.{constantDelay, limitRetriesByCumulativeDelay}
+import retry.retryingOnFailures
+
+object SelectablePeerDiscoveryDelay {
+
+  def make[F[_]: Async](
+    clusterStorage: ClusterStorage[F],
+    appEnvironment: AppEnvironment
+  ): PeerDiscoveryDelay[F] = new PeerDiscoveryDelay[F] {
+
+    private val logger = Slf4jLogger.getLogger[F]
+
+    private val checkPeersAttemptDelay = 1.minute
+    private val checkPeersMaxDelay = 10.minutes
+    private val additionalDiscoveryDelay = 10.minutes
+
+    private val minPeersCount: PosInt = 2
+    private val fallbackPeersCount: PosInt = 1
+
+    private def waitForPeers(minPeers: PosInt): F[Set[Peer]] = retryingOnFailures[Set[Peer]](
+      limitRetriesByCumulativeDelay(
+        threshold = checkPeersMaxDelay,
+        policy = constantDelay(checkPeersAttemptDelay)
+      ),
+      a => (a.size >= minPeers).pure[F],
+      (a, details) => logger.info(s"Discovered ${a.size} selectable peers, waiting for more: $details")
+    )(clusterStorage.getResponsivePeers.map(_.filter(_.state === Ready)))
+
+    override def waitForPeers: F[Unit] = {
+      if (appEnvironment === Dev) {
+        waitForPeers(fallbackPeersCount)
+      } else {
+        waitForPeers(minPeersCount) >>
+          Async[F].sleep(additionalDiscoveryDelay) >>
+          waitForPeers(fallbackPeersCount)
```

Comment:
`waitForPeers` checks to see if the requested number of peers has been discovered every 1 minute. It retries failures up to 10 mins. So, if it finds the specified number of peers before the 10 minute mark, it continues. Otherwise, it keeps trying until 10 minutes have elapsed.

After looking for 2 peers, @ryle-goehausen asked that we have an additional delay to discover more peers. That's what `Async[F].sleep(additionalDiscoveryDelay)` is for.

After that, we double-check that there is at least on selectable peer (`waitForPeers(fallbackPeersCount)`) so that `PeerSelect` doesn't blow-up. After that, `PeerSelect` throws `NoPeersToSelect` if there are no selectable peers.

The `additionalDiscoveryDelay` is being reduced to 3 minutes after talking with @ryle-goehausen .

Reply:
Ok, why din't you add those 3 minutes to either first or second "wait for peers"? You'd wait the same amount of time but eod you avoid blocking

Reply:
`waitForPeers`, the private method, does not necessarily wait the full 10 minutes. It waits until the specified number of selectable peers is discovered or 10 minutes, whichever comes first. It checks for selectable peers every 1 minute. So, in theory, it can finish right away, but it won't delay longer than 10 minutes. The `additionalDiscoveryDelay` is meant to wait for the full period of time. I also realized that `waitForPeers` is the name of both the public and private method. I've renamed the private method to `waitUntil`.